### ISSUE TITLE
Performance regression starting from 7.8.0 when too many search requests closes before getting result

### ISSUE THREAD

**Issue by libbkmz** [elastic/elasticsearch#75316]
**Elasticsearch version** (`bin/elasticsearch --version`):
`7.10.2`

**Plugins installed**: []

**JVM version** (`java -version`):
```
openjdk 15.0.1 2020-10-20
OpenJDK Runtime Environment AdoptOpenJDK (build 15.0.1+9)
OpenJDK 64-Bit Server VM AdoptOpenJDK (build 15.0.1+9, mixed mode, sharing)
```

**OS version** (`uname -a` if on a Unix-like system):
`Linux HOSTNAME 3.10.0-1160.31.1.el7.x86_64 #1 SMP Thu Jun 10 13:32:12 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux`

**Description of the problem including expected versus actual behavior**:
After the upgrade of the ES from 7.6.2 to 7.10.2 one of our clusters become degraded a lot in terms of latency and throughput. After a big and long investigation, I [found the suspect](https://github.com/elastic/elasticsearch/commit/61e5350e770966172cd64430661eeb903bc94518). It looks like starting from 7.8.0, we have this performance regression, even in the latest 7.13.3.
Our cluster has a unique workload that is used mainly for testing purposes. We are evaluating ElasticSearch as a DB backed for our app, and to gather some performance metrics in the long run, we have a "shadow traffic," which has mirror requests from our production app to ES, but if the gateway can get a response faster from its main data source than from ES, it will close request, and Nginx shows that as 499 status code. As a side-effect, I can't use almost any /_cat API requests. This is the example:

```
$ > time curl -sS localhost:9200/_cat/tasks | awk '{print $1}' | sort | uniq -c
0.01user 0.15system 6:01.05elapsed 0%CPU (0avgtext+0avgdata 1232maxresident)k
0inputs+0outputs (0major+350minor)pagefaults 0swaps
 256591 cluster:admin/tasks/cancel
   2243 cluster:admin/tasks/cancel[n]
     23 cluster:monitor/nodes/info
     39 cluster:monitor/nodes/stats
      1 cluster:monitor/stats
     18 cluster:monitor/tasks/lists
     83 cluster:monitor/tasks/lists[n]
 121370 indices:data/read/msearch
 121370 indices:data/read/search
     10 indices:data/read/search[can_match]
      3 indices:data/read/search[phase/fetch/id]
    141 indices:data/read/search[phase/query]
   2243 internal:admin/tasks/ban
``` 

While cluster has a load of search requests - I can't get much from it. Latency skyrockets from 400ms(99 percentile) to 20seconds - 120seconds ( 99percentile ) 

**Steps to reproduce**:

I was able to write [simple script](https://gist.github.com/libbkmz/199891acb668f6fff6a677d47f29b85d) which can reproduce the issue, unfortunately, one active instance is not enough, and to simulate our real production shadow workload, I have had to run 5 instances to get stuck in [setBan method](https://github.com/elastic/elasticsearch/blob/v7.10.2/server/src/main/java/org/elasticsearch/tasks/TaskManager.java#L354). Also, I'm attaching java flame graph captured by async-profiler for this particular problem. [es_setban_bug.flamegraph.html.gz](https://github.com/elastic/elasticsearch/files/6812577/es_setban_bug.flamegraph.html.gz)

Moreover, I don't know how good our search query parameters, but this one is very intensive, and we have a lot of queries like this, and I'm able to reproduce the issue with such a simple script. 

Our cluster has 83 nodes in total, with 73 of data nodes, 3 master nodes, and 6 proxy nodes. However, we found some issues with our hardware, and we were using only 1 proxy node for more than 2 months. So, all the ES requests come to a single ES instance with flags `it` only on version 7.10.2. But using more than 1 proxy node doesn't help at all - the outcome is the same...

**Comment by DaveCTurner** [comment]
The flame graph shows spending ~25% of the total time collecting the children of the banned parent:

<img width="965" alt="image" src="https://user-images.githubusercontent.com/5058284/125584372-89f6c076-2344-400a-a63d-8894c45c208a.png">

A linear search through 250k cancellable tasks will take some time indeed.

**Comment by libbkmz** [comment]
> A linear search through 250k cancellable tasks will take some time indeed:

And this number grows...

When I run only one instance of the script, ES can "chew" it for a long period of time without an increase in latencies and the number of tasks. However, after adding couple more instances of the workload script, the length of the tasks queue goes to the sky. After I got these results,  I couldn't get any more output from /_cat/tasks even after 30 minutes of waiting by curl.... Only after I stopped my scripts, I got the results immediately...

**Comment by DaveCTurner** [comment]
> After I got these results, I couldn't get any more output from /_cat/tasks even after 30 minutes of waiting by curl

I suspect this is better (or at least different) in 7.13 thanks to  #71171.

**Comment by libbkmz** [comment]
My production setup consists of ~80 boxes with 2 CPUs and 16threads in each. 128GB RAM on each. So, I still have 5 threads for the MANAGEMENT pool.
I have upgraded the cluster from 7.10.2 to 7.13.3. And now it behaves even worse. After starting the workload script, I can't receive any /_cat API after I stopped the script even after 30 minutes. So, I can't share the _cat/tasks numbers with 7.13.3 ... 

The flame graph shows worse time consumption for `setBan`. [es_7.13.3.setBan.flamegraph.html.gz](https://github.com/elastic/elasticsearch/files/6831975/es_7.13.3.setBan.flamegraph.html.gz)

<img width="1680" alt="image" src="https://user-images.githubusercontent.com/1144960/125977124-4356fed4-0e18-4e98-84b3-ebc38806d72e.png">




Also, it looks like I get data loss on a different cluster because of the `setBan` bugs...
`Cannot send ban for tasks with the parent [...:...] to the node [...]`
